kubernetes/cache.go

```diff
@@ -0,0 +1,458 @@
+		refreshDuration: refreshDuration,
+		stopChan:        nil,
+		controllers:     initControllers(clientset, refreshDuration),
+		syncCount:       0,
+		maxSyncCount:    20, // Move this to config ? or this constant is good enough ?
+	}
+	registerErrorCallback(newControllerImpl.ErrorCallback)
+
+	return &newControllerImpl
+}
+
+func initControllers(clientset kube.Interface, refreshDuration time.Duration) map[string]cache.SharedIndexInformer {
+	sharedInformers := informers.NewSharedInformerFactory(clientset, refreshDuration)
+	controllers := make(map[string]cache.SharedIndexInformer)
+	controllers["Pod"] = sharedInformers.Core().V1().Pods().Informer()
+	controllers["ReplicationController"] = sharedInformers.Core().V1().ReplicationControllers().Informer()
+	controllers["Deployment"] = sharedInformers.Apps().V1beta1().Deployments().Informer()
+	controllers["ReplicaSet"] = sharedInformers.Apps().V1beta2().ReplicaSets().Informer()
+	controllers["StatefulSet"] = sharedInformers.Apps().V1beta2().StatefulSets().Informer()
+	controllers["Job"] = sharedInformers.Batch().V1().Jobs().Informer()
+	controllers["CronJob"] = sharedInformers.Batch().V1beta1().CronJobs().Informer()
+	controllers["Service"] = sharedInformers.Core().V1().Services().Informer()
+	controllers["Endpoints"] = sharedInformers.Core().V1().Endpoints().Informer()
+
+	return controllers
+}
+
+func (c *controllerImpl) Start() {
+	if c.stopChan == nil {
+		c.stopChan = make(chan struct{})
+		go c.run(c.stopChan)
+		log.Infof("K8S cache started")
+	} else {
+		log.Warningf("K8S cache is already running")
+	}
+}
+
+func (c *controllerImpl) run(stop <-chan struct{}) {
+	for _, cn := range c.controllers {
+		go cn.Run(stop)
+	}
+	<-stop
+	log.Infof("K8S cache stopped")
+}
+
+func (c *controllerImpl) HasSynced() bool {
+	if c.syncCount > c.maxSyncCount {
+		log.Errorf("Max attempts reached syncing cache. Error connecting to k8s API ")
+		c.Stop()
+		return false
+	}
+	hasSynced := true
+	for _, cn := range c.controllers {
+		hasSynced = hasSynced && cn.HasSynced()
+	}
+	if hasSynced {
+		c.syncCount = 0
+	} else {
+		c.syncCount = c.syncCount + 1
+	}
+	return hasSynced
+}
+
+func (c *controllerImpl) WaitForSync() bool {
+	return cache.WaitForCacheSync(c.stopChan, c.HasSynced)
+}
+
+func (c *controllerImpl) Stop() {
+	if c.stopChan != nil {
+		close(c.stopChan)
+		c.stopChan = nil
+	}
+}
+
+func (c *controllerImpl) ErrorCallback(err error) {
+	if c.isErrorState == false {
+		log.Warningf("Error callback received: %s", err)
+		c.lastErrorLock.Lock()
+		c.isErrorState = true
+		c.lastError = err
+		c.lastErrorLock.Unlock()
+		c.Stop()
+	}
+}
+
+func (c *controllerImpl) checkStateAndRetry() error {
+	if c.isErrorState == false {
+		return nil
+	}
+
+	// Retry of the cache is hold by one single goroutine
+	c.lastErrorLock.Lock()
+	if c.isErrorState == true {
+		// ping to check if backend is still unavailable (used namespace endpoint)
+		_, err := c.clientset.CoreV1().Namespaces().List(emptyListOptions)
```

Comment:
Is this the least expensive API call we could make? Rather than ask for the list of namespaces, is there some kind of "status" endpoint? Or something that has a more predictably quick response time?

Reply:
This is a one time operation to recover full cache when backend is down.
clientset.Interface doesn't map a generic status function, so I used CoreV1() as representative here.
Fetching namespaces sounds correct to check that API is good.